### PR TITLE
Initial outline of builtin support in JS-API

### PR DESCRIPTION
**Pull request by eqrion**
WIP, going to keep adding commits here. #35 is related.

So far I have:
  1. A new section defining 'Builtins'
  2. WebAssemblyCompileOptions WebIDL and threading it through compilation methods
  3. New slot on Module for builtins

I've left some TODOs to continue developing.

cc @ajklein

### CODE REVIEW

**Comment by eqrion** [comment]
Okay, I think I just about have all the 'infrastructure' for builtin functions now. I still need to handle importedStringConstants. Then it's actually down to actually specifying each builtin function's behavior.

File: document/js-api/index.bs [hunk 1]
```diff
@@ -335,53 +341,102 @@ Note:
...
+To <dfn>instantiate imported strings</dfn> with module |module| and |importedStringModule|, perform the following steps:
+1. Assert: |importedStringModule| is not null.
+1. Let |exportsObject| be [=!=] [$OrdinaryObjectCreate$](null).
+        1. If |externtype| is of the form [=external-type/func=] <var ignore>functype</var>,
+1. [=list/iterate|For each=] (|moduleName|, |componentName|, |externtype|) of [=module_imports=](|module|),
+    1. If |moduleName| does not equal |importedStringModule|, then [=iteration/continue=].
+    1. Let |stringConstant| be |componentName|.
+    1. Let |status| be [=!=] [$CreateDataProperty$](|exportsObject|, |stringConstant|, |stringConstant|).
+    1. Assert: |status| is true.
+1. Perform [=!=] [$SetIntegrityLevel$](|exportsObject|, `"frozen"`).
```
Comment (ajklein) [inline document/js-api/index.bs]: Why do we need to freeze this object? It shouldn't even be allocated, let alone exposed to author code.

**Comment by eqrion** [comment]
Couple open questions:
  1. Does 'importedStringConstants' deserve a section somewhere explaining the feature? Builtins have a section it could go into, but the imported string stuff is technically orthogonal.
  2. What to call 'sets' of builtins inside the spec. I chose 'builtin sets', but arguably 'builtin module' is a better name. It just has overlap with the JS builtin module stuff which may cause an issue.
  3. Should the builtin specifier when used in the JS-API include the wasm prefix? `{builtins: ['wasm:js-string']}` vs `{builtins: ['js-string']}`.
    - I think I prefer it to include the wasm prefix just in case we allow other things in the future.
    - The explainer and engines goes the other way and doesn't accept the wasm prefix.
    - I think I accidentally specced it to require the prefix.
  4. Should we disallow: `{ builtins: ['js-string'], importedStringConstants: 'js-string' }`?
    - Right now the rules give imported strings precedence and they will shadow any builtins.

File: document/js-api/index.bs [hunk 2]
```diff
@@ -1646,6 +1714,114 @@ They expose the same interface as native JavaScript errors like {{TypeError}} an
...
+1. Let |builtins| be the result of [=get the builtins for a builtin set=] |builtinSetName|
+1. Let |exportsObject| be [=!=] [$OrdinaryObjectCreate$](null).
+        1. If |externtype| is of the form [=external-type/func=] <var ignore>functype</var>,
+1. [=list/iterate|For each=] (|name|, |funcType|, |algorithm|) of |builtins|,
+    1. Let |funcaddr| be the result fo [=create a builtin function=] with |funcType| and |algorithm|
+    1. Let |func| be the result of creating [=a new Exported Function=] from |funcaddr|.
+    1. Let |value| be |func|.
+    1. Let |status| be [=!=] [$CreateDataProperty$](|exportsObject|, |name|, |value|).
+    1. Assert: |status| is true.
+1. Perform [=!=] [$SetIntegrityLevel$](|exportsObject|, `"frozen"`).
```
Comment (ajklein) [inline document/js-api/index.bs]: Same question here re: freezing

**Review by ajklein — APPROVED** [review]
Mostly looks fine, and I'm fine landing as-is, but some questions below about freezing the export objects.

File: document/js-api/index.bs (hunk 1 above)
Comment (eqrion) [reply document/js-api/index.bs]: Just a copy-paste leftover from taking this from the 'create exports object' code. I'll remove it.

**Pull request merged**